### DISCUSSION TITLE
Ruleset-specific skinning support with new "skinnable target" flow

### DISCUSSION THREAD

**Discussion by frenzibyte** [ppy/osu#20694] [Ideas] [answered]
This initially started at https://github.com/ppy/osu/pull/20579#issuecomment-1268023338 where I was exploring through ways to get the legacy combo counter display different implementation based on the ruleset, and having it done by an imtermediate component that's added to the skin hierarchy doesn't work well (see [reason](https://github.com/ppy/osu/pull/20579#issuecomment-1272120646)).

I've continued exploring through other possible ways that involve changing the skin lookup flow, and I've came up with two different approaches after several tries and going back and forth on the structure:
 - https://github.com/ppy/osu/compare/master...frenzibyte:osu:ruleset-skinning-approach-1: changes the skin serialisation structure to be separated by rulesets, and adds a separate `RulesetHUDComponents` target that provides different components based on the ruleset.
 - https://github.com/ppy/osu/compare/master...frenzibyte:osu:ruleset-skinning-approach-2: assigns each `ISkinnableDrawable` its corresponding ruleset, and uses the same `MainHUDComponents` target for ruleset-specific components as well (serialisation structure shouldn't have been changed as each `ISkinnableDrawable` is already assigned its ruleset, but that brings up an edge case, see [commit message](https://github.com/ppy/osu/commit/e26feb1e15708439f23ade4490c00db0b0bb94c4)).

Interested to hear feedback on this, cc @peppy @smoogipoo.

**Comment by smoogipoo** [comment]
Attaching rulesets to components is not the first thing that came to my mind. This would be saved to the skin's json, right?

It's not immediately obvious to me why `RulesetHUDComponents` has to exist, based on the first approach alone, I have two files.

**MainHUDComponents.json**
```json
{
  "AllComponents": [],
  "DrawableComponentInfo": {
    "": []
  }
}
```
**RulesetHUDComponents.json**
```json
{
  "AllComponents": [],
  "DrawableComponentInfo": {
    "osu": []
  }
}
```

This kinda makes sense, but it's not obvious to me why `MainHUDComponents` couldn't have the "osu" key instead. Perhaps I'm missing something, and I also can't test because it crashes when clicking the lighting toolbox item, but that's my preliminary thought.

**Reply by peppy** [reply]
Also reading this feedback brings a question to mind: what if a user wants certain elements to show only for a specific ruleset, but those elements haven't been *hard defined* to be from that ruleset.

My immediate thinking is that to keep the skin editor simple, each component would get a new checkbox which defaults to enabled and reads along the lines of "Display in all rulesets". Unchecking it would make it only apply to the ruleset currently being edited. For components which are only available in the current ruleset, this would be `false` and disabled.

And I guess this also leads me to think that the first path is likely the one we want to take? I haven't spent too long looking at each from a code level, mind you. Hopefully the last paragraph helps with forward direction?

**Reply by frenzibyte** [reply]
> This kinda makes sense, but it's not obvious to me why MainHUDComponents couldn't have the "osu" key instead.

It's not possible due to the nature of how components are loaded off JSON files, since they're all loaded to one bindable components list inside `SkinnableTargetContainer`, making them lose track of which component belongs to which ruleset. This is why storing the ruleset for each `ISkinnableDrawable` is necessary (second approach).

The first approach avoids that by having a second `SkinnableTargetContainer` which receives a ruleset and loads/saves all of its corresponding components using the given ruleset. This makes the ability of changing whether a drawable should display on a specific ruleset or not hard to achieve, but still potentially feasible with some sort of `ReplaceRuleset` method in `Skin`.

> Also reading this feedback brings a question to mind: what if a user wants certain elements to show only for a specific ruleset, but those elements haven't been hard defined to be from that ruleset.
>
> My immediate thinking is that to keep the skin editor simple, each component would get a new checkbox which defaults to enabled and reads along the lines of "Display in all rulesets". Unchecking it would make it only apply to the ruleset currently being edited. For components which are only available in the current ruleset, this would be false and disabled.

This is essentially where I was leaning with my second approach. The blueprint would write to `ISkinnableDrawable.RulesetName` using some UI control (checkbox/etc.) and `Skin` will correctly include the component inside the correct target info dictionary for serialisation.

**Reply by frenzibyte** [reply]
> ...and I also can't test because it crashes when clicking the lighting toolbox item, but that's my preliminary thought.

I didn't notice that item actually exists, but it should be fixed now.

**Comment by frenzibyte** [comment, marked as answer]
Tracking via https://github.com/ppy/osu/issues/22388
